### PR TITLE
test: add lake benchmarks

### PR DESCRIPTION
**Pull request by tydeu**
@Kha Hopefully I configured this right.

### CODE REVIEW

**Comment by tydeu** [comment]
!bench

**Comment by tydeu** [comment]
!bench

**Comment by tydeu** [comment]
!bench

**Comment by tydeu** [comment]
!bench

**Comment by tydeu** [comment]
@Kha Any suggestion for improvements? Two potential uncovered test cases I thought of:

1. Multi-package elaboration
2. Modules with actual code

Case (2) could be reasonably addressed by benchmarking a build of the Lean core itself. However, I am not sure either of these cases would provide significantly different benchmarks to be worth the added complexity.

**Comment by tydeu** [comment]
I liked the benchmarks from #2463, so I went ahead and merged them into here.

**Comment by tydeu** [comment]
!bench

**Comment by tydeu** [comment]
Interesting, it appears a no-op build of Lean is faster than I no-op build of the inundation dummy project. So I guess number of files to hash is more significant than the contents of them?

**Review by Kha — CHANGES_REQUESTED** [review]
> Modules with actual code

Why is that interesting? We want to benchmark Lake, not Lean. A synthetic benchmark is great, it ensures that the input stays constant and cannot influence the timings.

**Comment by Kha** [comment]
> Modules with actual code

Ah, I suppose it is interesting e.g. in the context of #2444. I think I would still value isolated tests over that. It's okay to do benchmarks like this one manually if we don't expect output size to matter anymore after this change.

**Comment by tydeu** [comment]
@Kha Sime other interesting points of the Lean builds and real code builds:

 * Provides a time comparison with another build system (CMake/Make) -- we can officially say that, in parallel, Lake is faster than Make even for a complex project.

 * Enables measuring of code-level finer-grain dependency specification (like the forthcoming module system or leanprover/lake#86).

 * Similarly, things like asynchronous proof checking could also be captured by such a benchmark (if these are integrated with Lake enough to enable it to move onto building downstream files while the proofs are still being checked).

However, I am happy to remove the benchmark if you are confident it is not worth it. I was also initially unsure whether it was good idea.

**Comment by tydeu** [comment]
!bench

**Comment by tydeu** [comment]
!bench

**Comment by tydeu** [comment]
!bench

**Comment by tydeu** [comment]
!bench

File: tests/bench/inundation/README.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# Stress Test for Lake
+
+This folder generates a deeply nested import tree of Lean modules
+that can then be used as a stress test of large Lake builds (e.g., Mathlib).
+The the modules lack any code.
```
Comment (Kha) [inline tests/bench/inundation/README.md:5]:
```suggestion
The modules lack any code.

```

**Comment by tydeu** [comment]
> In any case, I'm still not convinced this is worth increasing the benchmark run time by 50%, no.

It may be worth noting that the Lean build itself is only contributing 4-5 minutes (so about 25% of the overall time), the other parts contribute about 1-3 minutes and benchmark without these changes is already usually 13-14 minutes. However, I presume that means you prefer me to take it out, so I will do that soon.

**Comment by Kha** [comment]
> It may be worth noting that the Lean build itself is only contributing 4-5 minutes (so about 25% of the overall time), the other parts contribute about 1-3 minutes and benchmark without these changes is already usually 13-14 minutes

I'm not sure what you are trying to tell me with that. All I see is a total increase from 13 minutes to 20 minutes.

**Pull request merged**